plugins/fs/src/file_path.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright 2019-2023 Tauri Programme within The Commons Conservancy
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: MIT
+
+use std::{
+    convert::Infallible,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+use serde::Serialize;
+use tauri::path::SafePathBuf;
+
+use crate::{Error, Result};
+
+/// Represents either a filesystem path or a URI pointing to a file
+/// such as `file://` URIs or Android `content://` URIs.
+#[derive(Debug, Serialize, Clone)]
+#[serde(untagged)]
+pub enum FilePath {
+    /// `file://` URIs or Android `content://` URIs.
+    Url(url::Url),
+    /// Regular [`PathBuf`]
+    Path(PathBuf),
+}
+
+/// Represents either a safe filesystem path or a URI pointing to a file
+/// such as `file://` URIs or Android `content://` URIs.
+#[derive(Debug, Clone, Serialize)]
+pub enum SafeFilePath {
+    /// `file://` URIs or Android `content://` URIs.
+    Url(url::Url),
+    /// Safe [`PathBuf`], see [`SafePathBuf``].
+    Path(SafePathBuf),
+}
+
+impl FilePath {
+    /// Get a reference to the contaiend [`Path`] if the variant is [`FilePath::Path`].
+    #[inline]
+    pub fn path(&self) -> Option<&Path> {
+        match self {
+            Self::Url(_) => None,
+            Self::Path(p) => Some(p),
+        }
+    }
+
+    /// Takes the contained [`PathBuf`] if the variant is [`FilePath::Path`],
+    /// and when possible, converts Windows UNC paths to regular paths.
+    #[inline]
+    pub fn simplified(self) -> Self {
+        match self {
+            Self::Url(url) => Self::Url(url),
+            Self::Path(p) => Self::Path(dunce::simplified(&p).to_path_buf()),
+        }
+    }
+
+    /// Try to convert into [`PathBuf`] if possible.
+    ///
+    /// This calls [`Url::to_file_path`](url::Url::to_file_path) if the variant is [`FilePath::Url`],
+    /// otherwise returns the contained [PathBuf] as is.
+    #[inline]
+    pub fn into_path(self) -> Result<PathBuf> {
+        match self {
+            Self::Url(url) => url
+                .to_file_path()
+                .map(PathBuf::from)
+                .map_err(|_| Error::InvalidPathUrl),
+            Self::Path(p) => Ok(p),
+        }
+    }
+}
+
+impl SafeFilePath {
+    /// Get a reference to the contaiend [`Path`] if the variant is [`FilePath::Path`].
+    #[inline]
+    pub fn path(&self) -> Option<&Path> {
```

Comment:
What about as_path to match PathBuf's api? https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.as_path

Reply:
makes sense